### ISSUE TITLE
Request: Example of Reducer for createAsyncAction

### ISSUE THREAD

**Issue by daveking999** [piotrwitek/typesafe-actions#68]
I would like to expand your Todo example by providing an action which fetches a list of Todo's from an API.

I've declared the following in ```actions.ts:```
```
const FETCH = 'todos/FETCH';
const FETCH_SUCCESS = 'todos/FETCH_SUCCESS';
const FETCH_FAILURE = 'todos/FETCH_FAILURE';

export const fetchTodos = createAsyncAction(
    FETCH,
    FETCH_SUCCESS,
    FETCH_FAILURE
)<void, Todo[], Error>();
```
and declared the API in ```root-epic.ts```

```
const fetchTodosFlow: Epic<RootAction, RootState> = (action$, store) =>
    action$.pipe(
        filter(isActionOf(fetchTodos.request)),
        mergeMap(action =>
            ajax.getJSON('http://someurl/fetch').pipe(
                map(fetchTodos.success),
                catchError(pipe(fetchTodos.failure, of))
            )
        ));
```
however I get the following tslint error:
```
Error:(14, 7) TS2322: Type '(action$: ActionsObservable<RootAction>, store: StateObservable<{ router: RouterState; todos: Readonly<{ todos: Todo[]; todosFetch: Todo[]; todosFilter: TodosFilter; }>; }>) => Observable<{} | PayloadAction<...>>' is not assignable to type 'Epic<RootAction, { router: RouterState; todos: Readonly<{ todos: Todo[]; todosFetch: Todo[]; todosFilter: TodosFilter; }>; }, any, RootAction>'.
  Type 'Observable<{} | PayloadAction<"todos/FETCH_SUCCESS", Todo[]>>' is not assignable to type 'Observable<RootAction>'.
    Type '{} | PayloadAction<"todos/FETCH_SUCCESS", Todo[]>' is not assignable to type 'RootAction'.
      Type '{}' is not assignable to type 'RootAction'.
        Type '{}' is not assignable to type 'LocationChangeAction'.
          Property 'type' is missing in type '{}'.
```

I'm using the following versions:
```
    "react-redux": "5.0.7",
    "react-router-redux": "5.0.0-alpha.9",
    "redux": "4.0.0",
    "redux-observable": "^1.0.0",
    "redux-thunk": "^2.3.0",
    "rxjs": "6.1.0",
    "rxjs-compat": "6.1.0",
```

Any help is greatly appreciated.

**Comment by piotrwitek** [comment]
Hey, that's cool, I'll try to book some time tomorrow and take a look.

**Comment by MattTennison** [comment]
I had the same issue. It's because the version of redux-observable in the todo-app is `v1.0.0-alpha-2` whereas you're using the latest version of redux-observable, `v1.0.0`. 

The Epic Signature has changed between those two releases: https://github.com/redux-observable/redux-observable/commit/0b0efc05ed7fb6f70efa3faa0e106a9a4384069c

So, you probably want:
`const fetchTodosFlow: Epic<RootAction, RootAction, RootState> = ...`, allowing you to output any action in your app in the output observable.

**Comment by daveking999** [comment]
Hi @MattTennison , I've tried out your suggestion but unfortunately it didn't work.
I downgraded RxJS to 5.x, **still no joy**.
I checked out the latest master of the todo-sample and tried it with no package upgrades, **even less joy :-(**

Can you push a branch of the sample app that you have working ?

Thanks again,
Dave

**Comment by MattTennison** [comment]
By todo-sample, do you mean this repository: https://github.com/piotrwitek/typesafe-actions-todo-app? Because it doesn't have any Epic's created in it - if you're struggling to get it running, perhaps there's an issue elsewhere.

I'm not too familiar with all this stuff myself, so can't provide much more help, but I can share my working Epic:

```
export type RootAction = ActionType<typeof Locations> | ActionType<typeof SearchBar>
export type RootState = StateType<typeof appReducer>;
```
```
export interface IServices {
  geolocationService: IGeolocationService;
  locationService: ILocationService;
}
```

```
export const searchBarFlow: Epic<RootAction, RootAction, RootState, IServices> = ((action$, state$, { locationService }) => {
  return action$.pipe(
    filter(isSearchRequest),
    switchMap(action =>
      from(performLocationSearch()).pipe(
        map(handleLocationSearchResponse),
        catchError(err => of(searchFromSearchBar.failure(err)))
      ))
  );

  function isSearchRequest(action: RootAction) {
    return action.type === getType(searchFromSearchBar.request);
  }

  function performLocationSearch() {
    return locationService.getNearbyLocationsObservable(state$.value.userInterface.searchBar.text);
  }

  function handleLocationSearchResponse(response: ILocationServiceResult<ILocationSearch>) {
    return response.isSuccessful ?
      searchFromSearchBar.success(response) :
      searchFromSearchBar.failure(new Error(response.error || "Something went wrong"));
  }
});
```
Hope this helps!

**Comment by IssuehuntBot** [comment]
@boostio funded this issue with $20. [Visit this issue on Issuehunt](https://issuehunt.io/repos/110746954/issues/68)

**Comment by IssuehuntBot** [comment]
@loadbalance-sudachi-kun funded this issue with $256. [Visit this issue on Issuehunt](https://issuehunt.io/repos/110746954/issues/68)

**Comment by IssuehuntBot** [comment]
@piotrwitek has started working. [Visit this issue on Issuehunt](https://issuehunt.io/repos/110746954/issues/68)

**Comment by Carl-Foster** [comment]
Is this not an issue with the catchError syntax? For our app, we always follow this:
```ts
ajax.getJson().pipe(
  map(successAction),
  catchError(err => of(failureAction(err))),
)
```

**Comment by denyo** [comment]
In addition to the example of @MattTennison what was missing for me (and not really documented anywhere) is where to get the service from within the epic.

```
export interface Services {
    positionService: PositionService;
}
```

The epic itself
```
export const fetchPositionsEpic: Epic<RootAction, RootAction, RootState, Services> = (action$, state$, { positionService }) =>
    action$.pipe(
        filter(isActionOf(fetchPositions.request)),
        switchMap(() =>
            from(positionService.getPositions()).pipe(
                map((data) => fetchPositions.success(data)),
                catchError((err) => of(fetchPositions.failure(err)))
            )
        )
    );
```

Then hooking it up
```
const rootEpic = combineEpics(fetchPositionsEpic);

// inside configureStore
const epicMiddleware = createEpicMiddleware({ dependencies: { positionService: new PositionService() } });

const store = createStore(
    rootReducer,
    initialState,
    composeEnhancers(applyMiddleware(...[thunk].concat(...middleware, epicMiddleware)))
);

epicMiddleware.run(rootEpic);
```

**Comment by piotrwitek** [comment]
Working example is there in the referenced commit, but docs examples still needs to be updated. WIP

**Comment by am17torres** [comment]
I'm not sure if I'm doing something obviously wrong but I have a similar problem (I think)

user.actions.ts
```typescript
import { createAsyncAction, ActionType } from "typesafe-actions";

import { User } from "./user.models";

const SIGN_IN = "@user/SIGN_IN";
const SIGN_IN_SUCCESS = "@user/SIGN_IN_SUCCESS";
const SIGN_IN_FAILURE = "@user/SIGN_IN_FAILURE";

export const signInAction = createAsyncAction(
  SIGN_IN,
  SIGN_IN_SUCCESS,
  SIGN_IN_FAILURE
)<void, User, Error>();

export type UserAction = ActionType<typeof signInAction>;
```

user.epic.ts
```typescript
import { combineEpics, Epic } from "redux-observable";
import { from, of } from "rxjs";
import { filter, switchMap, catchError, map } from "rxjs/operators";
import { isActionOf } from "typesafe-actions";

import { RootState } from "../../store/root-reducer";
import * as Services from "../../services";

import { UserAction, signInAction } from "./user.actions";

const signIn: Epic<UserAction, UserAction, RootState, typeof Services> = (
  action$,
  store,
  { GoogleAuthServices }
) => {
  action$.pipe(
    filter(isActionOf(signInAction.request)),
    switchMap(action =>
      from(GoogleAuthServices.signIn()).pipe(
        map(user => signInAction.success(user)),
        catchError(err => of(signInAction.failure(err)))
      )
    )
  );
};

export default combineEpics(signIn);
```

GoogleAuthService
```typescript
export class GoogleAuthServices {
  public static async signIn(): Promise<User> {
    const auth = await GoogleAuthServices.getAuth();
    return auth.signIn().then(
      (GoogleUser: any): User => {
        const profile = GoogleUser.getBasicProfile();
        return {
          name: profile.getName(),
          imageUrl: profile.getImageUrl()
        } as User;
      },
      (error: Error) => {
        throw error;
      }
    );
  }
}
```

Error 
```typescript
Type error: Type '(action$: ActionsObservable<EmptyAction<"@user/SIGN_IN"> | PayloadAction<"@user/SIGN_IN_SUCCESS",User> | PayloadAction<"@user/SIGN_IN_FAILURE", Error>>, store: StateObservable<{ user: Readonly<...>; }>, { GoogleAuthServices }: typeof import("../../services/index")) => void' is not assignable to type 'Epic<EmptyAction<"@user/SIGN_IN"> | PayloadAction<"@user/SIGN_IN_SUCCESS", User> | PayloadAction<"@user/SIGN_IN_FAILURE", Error>, EmptyAction<"@user/SIGN_IN"> | PayloadAction<...> | PayloadAction<...>, { ...; }, typeof import("../../services/index")>'.
  Type 'void' is not assignable to type 'Observable<EmptyAction<"@user/SIGN_IN"> | PayloadAction<"@user/SIGN_IN_SUCCESS", User> | PayloadAction<"@user/SIGN_IN_FAILURE", Error>>'.  TS2322
```

**Comment by IssuehuntBot** [comment]
@piotrwitek has submitted a pull request. [See it on IssueHunt](https://issuehunt.io/repos/110746954/issues/68)

**Comment by IssuehuntBot** [comment]
@piotrwitek has rewarded $193.20 to @piotrwitek. [See it on IssueHunt](https://issuehunt.io/repos/110746954/issues/68)

- :moneybag: Total deposit: $276.00
- :tada: Repository reward(20%): $55.20
- :wrench: Service fee(10%): $27.60
